src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/ECDsaCng.cs

```diff
@@ -1,157 +1,156 @@
+using Internal.Cryptography;
 
 namespace System.Security.Cryptography
 {
-    public partial class ECDsa : ECAlgorithm
+    public sealed partial class ECDsaCng : ECDsa
     {
+        private CngAlgorithmCore _core = new CngAlgorithmCore(nameof(ECDsaCng));
+        private CngAlgorithm _hashAlgorithm = CngAlgorithm.Sha256;
+
         /// <summary>
-        /// Creates an instance of the platform specific implementation of the cref="ECDsa" algorithm.
+        ///     Hash algorithm to use when generating a signature over arbitrary data
         /// </summary>
-        public static new partial ECDsa Create()
+        public CngAlgorithm HashAlgorithm
         {
-            return new ECDsaImplementation.ECDsaCng();
+            get
+            {
+                return _hashAlgorithm;
+            }
+            set
+            {
+                _hashAlgorithm = value ?? throw new ArgumentNullException(nameof(value));
+            }
         }
 
         /// <summary>
-        /// Creates an instance of the platform specific implementation of the cref="ECDsa" algorithm.
+        ///     Creates a new ECDsaCng object that will use the specified key. The key's
+        ///     <see cref="CngKey.AlgorithmGroup" /> must be ECDsa. This constructor
+        ///     creates a copy of the key. Hence, the caller can safely dispose of the
+        ///     passed in key and continue using the ECDsaCng object.
         /// </summary>
-        /// <param name="curve">
-        /// The <see cref="ECCurve"/> representing the elliptic curve.
-        /// </param>
-        public static partial ECDsa Create(ECCurve curve)
+        /// <param name="key">Key to use for ECDsa operations</param>
+        /// <exception cref="ArgumentException">if <paramref name="key" /> is not an ECDsa key</exception>
+        /// <exception cref="ArgumentNullException">if <paramref name="key" /> is null.</exception>
+        [SupportedOSPlatform("windows")]
+        public ECDsaCng(CngKey key)
         {
-            return new ECDsaImplementation.ECDsaCng(curve);
+            if (key == null)
+                throw new ArgumentNullException(nameof(key));
+
+            if (!IsEccAlgorithmGroup(key.AlgorithmGroup))
+                throw new ArgumentException(SR.Cryptography_ArgECDsaRequiresECDsaKey, nameof(key));
+
+            Key = CngAlgorithmCore.Duplicate(key);
         }
 
-        /// <summary>
-        /// Creates an instance of the platform specific implementation of the cref="ECDsa" algorithm.
-        /// </summary>
-        /// <param name="parameters">
-        /// The <see cref="ECParameters"/> representing the elliptic curve parameters.
-        /// </param>
-        public static partial ECDsa Create(ECParameters parameters)
+        protected override void Dispose(bool disposing)
         {
-            ECDsa ec = new ECDsaImplementation.ECDsaCng();
-            ec.ImportParameters(parameters);
-            return ec;
+            _core.Dispose();
```

Comment:
Should this be inside `if (disposing)`? Is finalizer somewhere going to mean `_core.Dispose()` going to get disposed twice?

Reply:
For sanity/consistency, probably.  But since the Cng classes are sealed, there's not a sneaky place a finalizer can come in.  While it looks like I massively rewrote this file, it's really "move the one from the other project over top this", where the Dispose already looked like that.